components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/EntitlementUtil.java

```diff
@@ -599,4 +601,33 @@ public static boolean isPolicyMetadataStoringEnabled() {
         // The default behavior is to store policy meta data.
         return StringUtils.isEmpty(propertyValue) || Boolean.parseBoolean(propertyValue);
     }
+
+    public static Map<String, Set<AttributeDTO>> getAttributesFromPolicies(PolicyDTO[] policyDTOs) {
+
+        Map<String, Set<AttributeDTO>> attributeMap = new HashMap<>();
+        for (PolicyDTO policyDTO : policyDTOs) {
+            Set<AttributeDTO> attributeDTOs = new HashSet<>(Arrays.asList(policyDTO.getAttributeDTOs()));
+            String[] policyIdRef = policyDTO.getPolicyIdReferences();
+            String[] policySetIdRef = policyDTO.getPolicySetIdReferences();
+
+            if (policyIdRef != null && policyIdRef.length > 0 || policySetIdRef != null && policySetIdRef.length > 0) {
+                for (PolicyDTO dto : policyDTOs) {
```

Comment:
The method creates a map of policy id : set of own policy attributes as well as referenced policy attributes
For a given policy (first loop), it gets the referenced policies, and using a second for loop identifies this referenced policy to add those attributes to the set as well. 
Since this is a legacy code, will not try to optimize.